### PR TITLE
[AC-4900] Update PATCH and POST for User

### PR DESCRIPTION
**Pull request by mc-nathan**
To test:
- `make load-remote-db DB_FILE_NAME=initial_schema.sql.gz`
- `make grant-permissions PERMISSION_USER=demoadmin@masschallenge.org PERMISSION_CLASSES=v1_clients,v1_confidential`
- `make dev`
- Login as demoadmin@masschallenge.org
- Do a POST to /v1/user with various parameters as described in the ticket.  See that you get what you expect.
- Do various PATCHes to different fields of /v1/user/<id> and see that it does what you expect.
- There should a reasonable level of error checking on the various fields.

Note: The PATCH related work is relatively well organized with appropriately shared validators and a shared setting mechanism.  The POST work leverages some of this, but could definitely be better integrated.  However, the OPTIONS are still haphazard.  Improvements to OPTIONS are planned in [AC-4857](https://masschallenge.atlassian.net/browse/AC-4857) as is a more general re-architecture that will bring GET, POST, PATCH and OPTIONS much closer together.  Given that this is a timeboxed ticket, we will push back on any requests to refactor POST to leverage more of the PATCH infrastructure.

### CODE REVIEW

File: web/impact/impact/v1/views/user_list_view.py [hunk 1]
```diff
@@ -62,20 +70,25 @@ def _copy_translated_keys(self, user_data, keys):
...
 
     def _profile_args(self, user_data):
-        self._check_required_keys(user_data, ProfileHelper.REQUIRED_KEYS)
+        self._check_required_keys(user_data, ProfileHelper.CORE_REQUIRED_KEYS)
         results = self._copy_translated_keys(user_data,
                                              ProfileHelper.INPUT_KEYS)
-        results["gender"] = self._find_gender(results.get("gender"))
+        user_type = validate_choices(
+            self, "user_type", results.get("user_type"), VALID_USER_TYPES)
+        if user_type == "expert":
```
Comment (yotammanor) [inline web/impact/impact/v1/views/user_list_view.py]: a named CONSTANT?

**Review by yotammanor — CHANGES_REQUESTED** [review]
Functional issues:

- in a POST, the boolean fields on Expert (speaker_interest, mentor_interest, is_active, judge_interest, office_hours_interest) expect python style True/False instead of the standard JS/JSON true/false (which is also the form returned in a GET. - **Fixed**

- further more, when passed an invalid value (currently "true/false" but any other, the error is an internal 500 error instead of a validation error displayed nicely to the user (and the error doesn't say which field is the bad one, which for the user is important). The validation can be postponed, but the correct values should be accepted. PATCH worked correctly. - **Fixed**

- URL fields should be valid for an empty string (DB has no `null=True` but, allows for empty strings, and there are many instances as such). Note that the current implementation allows add/update the url field, but does not allow removing it. Also, consider removing the URL validation altogether from linked_in_url and facebook_url since there's no such validation in the model itself. - **Fixed.  Validation issue for URLs was actually a deficiency in the model definitions for impact-api.  URLField is used in mc.models.CoreProfile and is now used in the corresponding models in impact-api in this PR.**

- successful PATCH returns an empty body. Is that an expected behavior? - **Yes. What else would you consider returning other than a 200?**

- When a field is invalid, the error message is duplicated multiple times if other fields are passed in, regardless of the other fields being valid: **Good catch!  FIxed.**

```

# Request:
wget --quiet \
  --method PATCH \
  --header 'content-type: application/x-www-form-urlencoded' \
  --header 'accept: application/json; indent=4' \
  --header 'authorization: Bearer HDaRMuL2mr51UVrpSAISZyN6eO6W7w' \
  --header 'cache-control: no-cache' \
  --header 'postman-token: 55fdee93-c7b4-69b3-69fa-195b0844e948' \
  --body-data expert_category=ff \
  --output-document \
  - http://localhost:8000/api/v1/user/47672/

# Response:
[
    "Invalid expert_category: Expected 'Executive', 'Investor', 'Lawyer' or 'Other' not ff"
]

# Request 2:
wget --quiet \
  --method PATCH \
  --header 'content-type: application/x-www-form-urlencoded' \
  --header 'accept: application/json; indent=4' \
  --header 'authorization: Bearer HDaRMuL2mr51UVrpSAISZyN6eO6W7w' \
  --header 'cache-control: no-cache' \
  --header 'postman-token: 709a6896-4a4d-c82a-5142-c8ce3febd04f' \
  --body-data 'expert_category=ff&speaker_interest=True' \
  --output-document \
  - http://localhost:8000/api/v1/user/47672/

# Response 2:
[
    "Invalid expert_category: Expected 'Executive', 'Investor', 'Lawyer' or 'Other' not ff",
    "Invalid expert_category: Expected 'Executive', 'Investor', 'Lawyer' or 'Other' not ff"
]
```

- in a POST, I tried to create an entrepreneur with expert only parameters. I got some inconsistent validation error messages:
```
[
    "'judge_interest' is not supported for entrepreneur",
    "'home_program_family_id' is not supported for entrepreneur",
    "'company' is not supported for entrepreneur",
    "'title' is not supported for entrepreneur",
    "'mentor_interest' is not supported for entrepreneur",
    "'primary_industry_id' is not supported for entrepreneur",
    "'expert_category' is not supported for entrepreneur",
    "'office_hours_interest' is not supported for entrepreneur",
    "'speaker_interest' is not supported for entrepreneur",
    "Valid keys are: ['email', 'first_name', 'last_name', 'is_active', 'gender', 'user_type', 'company', 'expert_category', 'home_program_family_id', 'phone', 'primary_industry_id', 'title', 'judge_interest', 'personal_website_url', 'bio', 'twitter_handle', 'office_hours_interest', 'referred_by', 'office_hours_topics', 'speaker_topics', 'linked_in_url', 'phone', 'speaker_interest', 'facebook_url', 'mentor_interest']"
]
```
**Scratched my head a bit on this, but I think what you are referring to is the inaccuracy of the list of valid keys in the last message.  I've corrected that.**

- Trying to POST a member with a _referred_by, office_hours_topics, speaker_topics_ fields raises a 500 (instead of being identified as invalid keys for a member, like other such keys). **Fixed**

- Trying to submit a POST with a member with a "bio" field returns a 500 instead of a validation error. **Fixed**

- In definition of DONE it says `"first_name": field is different from key` - I don't understand what this requires, but I also don't see any related validation code for this field in the PR, so I'm suspicious.   **Technically that's in the roadmap.  The point there is that the storage key is "last_name" for "first_name" and "short_name" for "last_name" due to the way SimpleUser is defined.  These were just nots to make sure that the right thing was happening.**

style and coverage are good. A few code related comments below, but the focus should mostly be the functional part.

**Comment by yotammanor** [comment]
@mc-nathan note that AC-4947 is here as well as in the separate PR (#49).
Not sure if it's on purpose or not, but pointing it out either way.

File: web/impact/impact/tests/test_user_detail_view.py [hunk 2]
```diff
@@ -24,48 +31,233 @@ def test_get(self):
...
+            assert helper.field_value("bio") == bio
+            assert helper.field_value("facebook_url") == facebook_url
+            assert helper.field_value("gender") == "m"
+            assert helper.field_value("linked_in_url") == linked_in_url
+            assert helper.field_value("personal_website_url") == website_url
+            assert helper.field_value("phone") == phone
+            assert helper.field_value("twitter_handle") == twitter_handle
+
+    def test_patch_expert_fields(self):
+        context = UserContext(user_type="EXPERT")
```
[the remark this answers is not in the document]
Comment (mc-nathan) [reply web/impact/impact/tests/test_user_detail_view.py]: Done

File: web/impact/impact/tests/test_user_detail_view.py [hunk 3]
```diff
@@ -24,48 +31,233 @@ def test_get(self):
...
+            assert (helper.field_value("office_hours_topics") ==
+                    office_hours_topics)
+            assert helper.field_value("referred_by") == referred_by
+            assert helper.field_value("speaker_interest") is True
+            assert helper.field_value("speaker_topics") == speaker_topics
+            assert helper.field_value("judge_interest") is False
+            assert helper.field_value("mentor_interest") is True
+
+    def test_patch_personal_website_url_with_username_and_password_url(self):
+        context = UserContext(user_type="ENTREPRENEUR")
```
[the remark this answers is not in the document]
Comment (mc-nathan) [reply web/impact/impact/tests/test_user_detail_view.py]: Done

File: web/impact/impact/tests/test_user_detail_view.py [hunk 4]
```diff
@@ -24,48 +31,233 @@ def test_get(self):
...
+                "personal_website_url": website_url,
+                }
+            self.client.patch(url, data)
+            user.refresh_from_db()
+            profile.refresh_from_db()
+            helper = UserHelper(user)
+            assert helper.field_value("personal_website_url") == website_url
+
+    def test_patch_expert_field_fails_for_entrepreneur(self):
+        context = UserContext(user_type="ENTREPRENEUR")
```
[the remark this answers is not in the document]
Comment (mc-nathan) [reply web/impact/impact/tests/test_user_detail_view.py]: Done

File: web/impact/impact/tests/test_user_detail_view.py [hunk 5]
```diff
@@ -24,48 +31,233 @@ def test_get(self):
...
+            url = reverse("user_detail", args=[user.id])
+            bad_value = "This is *not* a valid url"
+            response = self.client.patch(url,
+                                         {"personal_website_url": bad_value})
+            assert response.status_code == 403
+            assert any(bad_value in datum
+                       for datum in response.data)
+
+    def test_patch_ids(self):
+        context = UserContext(user_type="EXPERT")
```
[the remark this answers is not in the document]
Comment (mc-nathan) [reply web/impact/impact/tests/test_user_detail_view.py]: Done

File: web/impact/impact/tests/test_user_detail_view.py [hunk 6]
```diff
@@ -24,48 +31,233 @@ def test_get(self):
...
+                    })
+            user.refresh_from_db()
+            profile.refresh_from_db()
+            helper = UserHelper(user)
+            assert (helper.field_value("home_program_family_id") ==
+                    program_family.id)
+            assert helper.field_value("primary_industry_id") == industry.id
+
+    def test_patch_invalid_primary_industry_id(self):
+        context = UserContext(user_type="EXPERT")
```
[the remark this answers is not in the document]
Comment (mc-nathan) [reply web/impact/impact/tests/test_user_detail_view.py]: Done

File: web/impact/impact/v1/helpers/model_helper.py [hunk 7]
```diff
@@ -17,6 +37,96 @@ def field_value(self, field):
...
+    return value
+
+
+def validate_choices(helper, field, value, choices, translations={}):
+    validate_string(helper, field, value)
+    result = translations.get(value, value)
+    if value in choices or result in choices:
+        return result
+    if isinstance(result, str):
+        result = translations.get(result.lower(), result)
```
[the remark this answers is not in the document]
Comment (mc-nathan) [reply web/impact/impact/v1/helpers/model_helper.py:92]: Because value isn't necessarily a string and I only wanted to try the .lower() if nothing else worked.  This is needed for "gender" so that "Male" gets properly translated to "m".

File: web/impact/impact/v1/helpers/profile_helper.py [hunk 8]
```diff
@@ -7,44 +32,203 @@
...
+]
+
+INVALID_INDUSTRY_ID_ERROR = ("Invalid {field}: "
+                             "Expected valid id for an industry resource")
+INVALID_PROGRAM_FAMILY_ID_ERROR = (
+    "Invalid {field}: Expected valid id for an program family resource")
+
+PHONE_REGEX = re.compile(r'^[0-9x.+() -]+$')
+TWITTER_REGEX = re.compile(r'^\S+$')
+# EMAIL_REGEX = re.compile(r'^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$')
```
[the remark this answers is not in the document]
Comment (mc-nathan) [reply web/impact/impact/v1/helpers/profile_helper.py]: Deleted.

File: web/impact/impact/v1/views/user_list_view.py (hunk 1 above)
Comment (mc-nathan) [reply web/impact/impact/v1/views/user_list_view.py]: Done.

**Comment by mc-nathan** [comment]
@yotam I believe I've addressed all your comments.  Thanks for the through review!

**Comment by yotammanor** [comment]
@mc-nathan  functional follow up:

- in a POST, the boolean fields on Expert (speaker_interest, mentor_interest, is_active, judge_interest, office_hours_interest) expect python style True/False instead of the standard JS/JSON true/false (which is also the form returned in a GET. - **Fixed** - field `is_active` is still not fixed. Others are good.

- successful PATCH returns an empty body. Is that an expected behavior? - **Yes. What else would you consider returning other than a 200**? - after a short google search on the subject, I now expect a 204 status that marks a no-body successful response ([reference](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PATCH)). Non blocking.

- in a POST, I tried to create an entrepreneur with expert only parameters. I got some inconsistent validation error messages: - **Scratched my head a bit on this, but I think what you are referring to is the inaccuracy of the list of valid keys in the last message. I've corrected that.** - Yes, that's what I was referring to. Sorry for not being clear about it.

- In definition of DONE it says "first_name": field is different from key - I don't understand what this requires, but I also don't see any related validation code for this field in the PR, so I'm suspicious. **Technically that's in the roadmap. The point there is that the storage key is "last_name" for "first_name" and "short_name" for "last_name" due to the way SimpleUser is defined. These were just nots to make sure that the right thing was happening.** - I've looked on the roadmap before, and I've read your answer now, and I've looked on the ticket (DONE, roadmap, comments), and I still don't understand the validation, so I don't know how to test this. I did manage to pass in first_name and last_names to POST and PATCH, so I assume this means that it's working fine (at least in a sanity test sense), but I did not understand what is the edge case to test against.

- Trying to POST a member with a referred_by, office_hours_topics, speaker_topics fields raises a 500 (instead of being identified as invalid keys for a member, like other such keys). **Fixed** - still seeing this issue, as is described.

**Also, another functional issue:**

- Trying to PATCH a non existing user raises a 500 instead of (I guess) 404.

- URL validation now works well in PATCH, but POST accepts invalid values.

**Review by yotammanor — CHANGES_REQUESTED** [review]
@mc-nathan  functional follow up:

- in a POST, the boolean fields on Expert (speaker_interest, mentor_interest, is_active, judge_interest, office_hours_interest) expect python style True/False instead of the standard JS/JSON true/false (which is also the form returned in a GET. - **Fixed** - _field `is_active` is still not fixed. Others are good._

- successful PATCH returns an empty body. Is that an expected behavior? - **Yes. What else would you consider returning other than a 200**? - _after a short google search on the subject, I now expect a 204 status that marks a no-body successful response ([reference](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PATCH)). Non blocking._

- in a POST, I tried to create an entrepreneur with expert only parameters. I got some inconsistent validation error messages: - **Scratched my head a bit on this, but I think what you are referring to is the inaccuracy of the list of valid keys in the last message. I've corrected that.** - _Yes, that's what I was referring to. Sorry for not being clear about it._

- In definition of DONE it says "first_name": field is different from key - I don't understand what this requires, but I also don't see any related validation code for this field in the PR, so I'm suspicious. **Technically that's in the roadmap. The point there is that the storage key is "last_name" for "first_name" and "short_name" for "last_name" due to the way SimpleUser is defined. These were just nots to make sure that the right thing was happening.** - _I've looked on the roadmap before, and I've read your answer now, and I've looked on the ticket (DONE, roadmap, comments), and I still don't understand the validation, so I don't know how to test this. I did manage to pass in first_name and last_names to POST and PATCH, so I assume this means that it's working fine (at least in a sanity test sense), but I did not understand what is the edge case to test against._

- Trying to POST a member with a referred_by, office_hours_topics, speaker_topics fields raises a 500 (instead of being identified as invalid keys for a member, like other such keys). **Fixed** _- still seeing this issue, as is described._

**Also, two more functional issues:**

- Trying to PATCH a non existing user raises a 500 instead of (I guess) 404.

- URL validation now works well in PATCH, but POST accepts invalid values.

**Comment by pbx** [comment]
Re the question of what a successful PATCH should return, I suggest "204 No Content". ["The server has fulfilled the request but does not need to return an entity-body"](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html)

**Comment by mc-nathan** [comment]
Addressed all of @yotam's latest concerns.  Successful PATCH is returning 204.  Provided full validation for POST in addition to PATCH which dealt with a lot of the issues.  Added better error handling for PATCH when the id is invalid.  Tests for everything.

File: web/impact/impact/v1/helpers/profile_helper.py [hunk 9]
```diff
@@ -7,58 +32,218 @@
...
+    if user_type not in user_types:
+        helper.errors.append(INVALID_CHOICE_ERROR.format(
+                field=field,
+                value=user_type,
+                choices=format_choices(user_types)))
+    return value
+
+
+def validate_personal_website_url(helper, field, value):
+    # This is essentially copied from mc.models.utils in accelerate.
```
Comment (pbx) [inline web/impact/impact/v1/helpers/profile_helper.py:92]: In the interest of sticking to our [open source standards](https://github.com/masschallenge/standards/blob/master/open_source_repos.md), we should either remove this entire comment block, or make a sanitized edit that doesn't reference MC, Accelerate, or JIRA ticket numbers.

File: web/impact/impact/v1/helpers/model_helper.py [hunk 10]
```diff
@@ -17,6 +37,97 @@ def field_value(self, field):
             return result
         return getattr(self.subject, field, None)
 
+    def field_setter(self, field, value):
```
Comment (pbx) [inline web/impact/impact/v1/helpers/model_helper.py:40]: Certainly not a blocker, but when I saw the name `field_setter` I guessed that a callable would be returned. Something like `set_field` seems like it would be more straightforward for this.

**Review by pbx — APPROVED** [review]
This works. I can't pretend to have been as exhaustive as Yotam, but I tested it to my satisfaction.  It was fun seeing PATCH in action. My comments are small; even the white-labeling one isn't worth being a blocker at this point.

**Review by yotammanor — APPROVED** [review]
